Applications/mnist/jni/main.cpp

```diff
@@ -0,0 +1,293 @@
+    l.resize(total_label_size);
+
+    getData(F, o, l, memI[i]);
+
+    for (unsigned int j = 0; j < feature_size; ++j)
+      outVec[i * feature_size + j] = o[j];
+    for (unsigned int j = 0; j < total_label_size; ++j)
+      outLabel[i * total_label_size + j] = l[j];
+  }
+
+  F.close();
+  return true;
+}
+
+/**
+ * @brief      get data which size is mini batch for validation
+ * @param[out] outVec
+ * @param[out] outLabel
+ * @param[out] status for error handling
+ * @retval true/false false : end of data
+ */
+bool getMiniBatch_val(float *outVec, float *outLabel, int *status) {
+
+  std::vector<int> memI;
+  std::vector<int> memJ;
+  unsigned int count = 0;
+  int data_size = total_val_data_size;
+
+  std::string filename = "mnist_trainingSet.dat";
+  std::ifstream F(filename, std::ios::in | std::ios::binary);
+
+  for (unsigned int i = 0; i < total_label_size * data_size; i++) {
+    if (!valduplicate[i])
+      count++;
+  }
+
+  if (count < mini_batch) {
+    for (unsigned int i = 0; i < total_label_size * data_size; ++i) {
+      valduplicate[i] = false;
+    }
+    return false;
+  }
+
+  count = 0;
+  while (count < mini_batch) {
+    int nomI = rangeRandom(0, total_label_size * data_size - 1);
+    if (!valduplicate[nomI]) {
+      memI.push_back(nomI);
+      valduplicate[nomI] = true;
+      count++;
+    }
+  }
+
+  for (unsigned int i = 0; i < count; i++) {
+    std::vector<float> o;
+    std::vector<float> l;
+
+    o.resize(feature_size);
+    l.resize(total_label_size);
+
+    getData(F, o, l, memI[i]);
+
+    for (unsigned int j = 0; j < feature_size; ++j)
+      outVec[i * feature_size + j] = o[j];
+    for (unsigned int j = 0; j < total_label_size; ++j)
+      outLabel[i * total_label_size + j] = l[j];
+  }
+
+  F.close();
+  return true;
+}
+
+
+/**
+ * @brief     create NN
+ *            Get Feature from tflite & run foword & back propatation
+ * @param[in]  arg 1 : configuration file path
+ * @param[in]  arg 2 : resource path
+ */
+int main(int argc, char *argv[]) {
+  if (argc < 2) {
+    std::cout << "./nntrainer_mnist mnist.ini\n";
+    exit(0);
+  }
+  const std::vector<std::string> args(argv + 1, argv + argc);
+  std::string config = args[0];
+
+  srand(time(NULL));
+  std::vector<std::vector<float>> inputVector, outputVector;
+  std::vector<std::vector<float>> inputValVector, outputValVector;
+  std::vector<std::vector<float>> inputTestVector, outputTestVector;
+
+  // This is to check duplication of data
+  for (unsigned int i = 0; i < total_label_size * total_train_data_size; ++i) {
+    duplicate[i] = false;
+  }
+
+  for (unsigned int i = 0; i < total_label_size * total_val_data_size; ++i) {
+    valduplicate[i] = false;
+  }
+
+  /**
+   * @brief     Neural Network Create & Initialization
+   */
+  nntrainer::NeuralNetwork NN(config);
+  NN.init();
+  NN.readModel();
+
+  /**
+   * @brief     Neural Network Train & validation
+   */
+  NN.train(getMiniBatch_train, getMiniBatch_val, nullptr);
```

Comment:
These functions `getMiniBatch_train/val` with same semantics are used everywhere. Do we want to provide these with our headers/C-API? If yes, will add.

Reply:
Well, this does not use c-api. We can rewrite this with C-API. As you can see, this uses cpp. if you are talking about C++API, then, yes. we definitely needs it.

Reply:
Fair enough. These can be removed with C++ API.